src/uct/ib/ud/base/ud_ep.c

```diff
@@ -684,6 +684,40 @@ static uct_ud_ep_t *uct_ud_ep_create_passive(uct_ud_iface_t *iface, uct_ud_ctl_h
     return ep;
 }
 
+static void uct_ud_ep_rx_ctl_drop_packet(uct_ud_ep_t *ep, uct_ud_neth_t *neth)
+{
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                           uct_ud_iface_t);
+    uct_ud_ctl_hdr_t *ctl = (uct_ud_ctl_hdr_t*)(neth + 1);
+
+    ucs_assert((ctl->type == UCT_UD_PACKET_CREQ) ||
+               (ctl->type == UCT_UD_PACKET_CREP));
+    ucs_trace_data("drop %s packet (neth_psn=%u) on iface=%p ep=%p conn_sn=%d"
+                   " ep_id=%d, dest_ep_id=%d rx_psn=%u ep_flags=0x%x"
+                   " ctl_ops=0x%x rx_creq_count=%d",
+                   (ctl->type == UCT_UD_PACKET_CREQ) ? "CREQ" : "CREP",
+                   neth->psn, iface, ep, ep->conn_sn, ep->ep_id,
+                   ep->dest_ep_id, ep->rx.ooo_pkts.head_sn, ep->flags,
+                   ep->tx.pending.ops, ep->rx_creq_count);
+
+    ucs_assertv_always((/* Drop CREQ, if the endpoint received duplicate
```

Comment:
1. I did my best to simplify as soon as I could :) IMO having one function reduces code duplication, since no need to write assertion and its printf's arguments two times.
2. I see, it was copied from the previous assertion.